docs: Updated and expanded alert documentation (#137)
paradedb/charts: charts/paradedb/docs/runbooks/CNPGClusterPhysicalReplicationLagCritical.md

CNPGClusterPhysicalReplicationLagCritical
=========================================

Meaning
-------

This alert is triggered when the physical replication lag of the CloudNativePG cluster exceeds `15s`.

Impact
------

> High physical replication lag can cause the cluster replicas to become out of sync. Queries to the `-r` and `-ro` endpoints may return stale data. In the event of a failover, the data that has not yet been replicate from the primary to the replicas may be lost.

Reviewer: ```suggestion
High physical replication lag can cause the cluster replicas to become out of sync. Queries to the `-r` and `-ro` endpoints may return stale data. In the event of a failover, the data that has not yet been replicated from the primary to the replicas may be lost.
```

Diagnosis
---------

You can identify the current primary instance with the [CloudNativePG Grafana Dashboard](https://grafana.com/grafana/dashboards/20417-cloudnativepg/) or via the following command:

```bash
kubectl exec --namespace <namespace> --stdin --tty services/<cluster_name>-rw -- psql -c "SELECT * FROM pg_stat_replication;"
```

High physical replication lag can be caused by a number of factors, including:

- Network issues and network congestion of the node network interface

> Check the network interface statistics using the Grafana Dashboard in the `Kubernetes Cluster` section.

Reviewer: ```suggestion
Check the network interface statistics using the `Kubernetes Cluster` section of the Grafana dashboard.
```

- High load on the primary or standby replicas

> Check the CPU and Memory usage of the CloudNativePG cluster instances using the [CloudNativePG Grafana Dashboard](https://grafana.com/grafana/dashboards/20417-cloudnativepg/).

Reviewer: ```suggestion
Inspect the CPU and Memory usage of the CloudNativePG cluster instances using the [CloudNativePG Grafana Dashboard](https://grafana.com/grafana/dashboards/20417-cloudnativepg/).
```

- Disk IO bottlenecks on the replicas

> Check the disk IO statistics using the [CloudNativePG Grafana Dashboard](https://grafana.com/grafana/dashboards/20417-cloudnativepg/).

Reviewer: ```suggestion
Inspect the disk IO statistics using the [CloudNativePG Grafana Dashboard](https://grafana.com/grafana/dashboards/20417-cloudnativepg/).
```

- Long-running queries

> Check the `Stat Activity` section of the [CloudNativePG Grafana Dashboard](https://grafana.com/grafana/dashboards/20417-cloudnativepg/).

Reviewer: ```suggestion
Inspect the `Stat Activity` section of the [CloudNativePG Grafana Dashboard](https://grafana.com/grafana/dashboards/20417-cloudnativepg/).
```

> - Suboptimal PostgreSQL configuration, in particular a small number of `max_wal_senders`. It should be set to a number greater than or equal to the number of instances in your cluster. It defaults to `10` so it is usually sufficient for a clusters with less than 10 instances.

Reviewer: ```suggestion
- Suboptimal PostgreSQL configuration, in particular a small number of `max_wal_senders`. It should be set to a number greater than or equal to the number of instances in your cluster. The default value of `10` is usually sufficient.
```

> You can check active PostgreSQL parameter configuration using the [CloudNativePG Grafana Dashboard](https://grafana.com/grafana/dashboards/20417-cloudnativepg/) in the `PostgreSQL Parameters` section.

Reviewer: ```suggestion
Inspect the `PostgreSQL Parameters` section of the [CloudNativePG Grafana Dashboard](https://grafana.com/grafana/dashboards/20417-cloudnativepg/).
```

Mitigation
----------

> - Kill any long-running transactions that could be creating more changes than standby replicas are able to process.

Reviewer: ```suggestion
- Kill any long-running transactions that could be creating more changes than the standby replicas can process.
```

```bash
kubectl exec -it services/paradedb-rw --namespace NAMESPACE -- psql
```

> - Increase the Memory and CPU resources of ParadeDB instances if they are under heavy load. You can do this by increasing the resource requests by setting `cluster.resources.requests` and `cluster.resources.limits` in your Helm values. It is highly recommended that you set both `requests` and `limits` to the same value to achieve QoS `Guaranteed`. This will require a restart of the CloudNativePG cluster instances and a primary switchover, which will cause a brief service disruption.

Reviewer: ```suggestion
- Increase the Memory and CPU resources of the ParadeDB instances under heavy load. This can be done by setting `cluster.resources.requests` and `cluster.resources.limits` in your Helm values. It is highly recommended to set both `requests` and `limits` to the same value to achieve QoS `Guaranteed`. This will require a restart of the CloudNativePG cluster instances and a primary switchover, which will cause a brief service disruption.
```

> If using the ParadeDB BYOC Terraform module, you can achieve the same thing by setting the `paradedb.cpu` and `paradedb.mem` parameters in the BYOC values.

Reviewer: ```suggestion
If using the ParadeDB BYOC Terraform module, this can be done by setting the `paradedb.cpu` and `paradedb.mem` parameters in the `.tfvars` file(s).
```

> - Enabling `wal_compression` by setting the `cluster.postgresql.parameters.wal_compression` parameter to `on`. might reduce the size of the WAL files and help reduce replication lag in a congested network. Changing `wal_compression` doesn't require a restart of the CloudNativePG cluster instances and normally can be done live.

Reviewer: ```suggestion
- Enable `wal_compression` by setting the `cluster.postgresql.parameters.wal_compression` parameter to `on`. Doing so will reduce the size of the WAL files and can help reduce replication lag in a congested network. Changing `wal_compression` does not require a restart of the CloudNativePG cluster.
```

> If you are using the ParadeDB BYOC Terraform module, you can set `paradedb.postgresql.parameters.wal_compression`.

Reviewer: ```suggestion
If using the ParadeDB BYOC Terraform module, this can be done by setting `paradedb.postgresql.parameters.wal_compression`.
```

> - Increasing the number of IOPS or throughput of the storage used by the CloudNativePG cluster instances can help reduce replication lag if disk IO bottlenecked. Doing that requires creating a new storage class with higher IOPS or throughput and rebuilding cluster instances one by one using the new storage class. This is a slow process that will also affect the cluster's availability.

Reviewer: ```suggestion
- Increase the number of IOPS/throughput of the storage used by the CloudNativePG cluster instances. Doing so can help reduce replication lag if the disk IO is bottlenecked. This requires creating a new storage class with higher IOPS/throughput and rebuilding cluster instances and their PVCs one by one using the new storage class. This is a slow process that will also affect the cluster's availability.
```
